### PR TITLE
ddl: support drop columns for multi-schema change

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/14766

### What is changed and how it works?

This PR supports dropping columns for multi-schema change.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/column.go [hunk 1]
```diff
@@ -641,33 +653,34 @@ func onDropColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
...
-		return nil, nil, nil, errors.Trace(err)
+		return nil, nil, nil, false, errors.Trace(err)
 	}
 
 	var colName model.CIStr
+	var ifExists bool
 	// indexIds is used to make sure we don't truncate args when decoding the rawArgs.
 	var indexIds []int64
-	err = job.DecodeArgs(&colName, &indexIds)
+	err = job.DecodeArgs(&colName, &ifExists, &indexIds)
```
Comment (wjhuang2016) [inline ddl/column.go:667]: Can it support rolling upgrade?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2978,13 +2978,24 @@ func checkMultiSpecs(sctx sessionctx.Context, specs []*ast.AlterTableSpec) error
...
 			return dbterror.ErrRunMultiSchemaChanges
 		}
 	}
 	return nil
 }
 
+func allSupported(specs []*ast.AlterTableSpec) bool {
+	for _, s := range specs {
+		switch s.Tp {
+		case ast.AlterTableAddColumns, ast.AlterTableDropColumn:
```
Comment (wjhuang2016) [inline ddl/ddl_api.go:2991]: DropColumn?

Comment (tangenta) [reply ddl/ddl_api.go:2991]: There is no definition of `ast.AlterTableDropColumns`.

File: ddl/column.go [hunk 3]
```diff
@@ -573,10 +573,21 @@ func checkDropColumnForStatePublic(tblInfo *model.TableInfo, colInfo *model.Colu
 }
 
 func onDropColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
-	tblInfo, colInfo, idxInfos, err := checkDropColumn(t, job)
+	tblInfo, colInfo, idxInfos, ifExists, err := checkDropColumn(t, job)
 	if err != nil {
+		if ifExists && dbterror.ErrCantDropFieldOrKey.Equal(err) {
```
Comment (wjhuang2016) [inline ddl/column.go:578]: Please add a comment for it. Why need to check ErrCantDropFieldOrKey, is it enough?

File: ddl/ddl_worker.go [hunk 4]
```diff
@@ -434,6 +434,11 @@ func (w *worker) deleteRange(ctx context.Context, job *model.Job) error {
 
 func jobNeedGC(job *model.Job) bool {
 	if !job.IsCancelled() {
+		if job.Warning != nil && dbterror.ErrCantDropFieldOrKey.Equal(job.Warning) {
```
Comment (wjhuang2016) [inline ddl/ddl_worker.go:437]: ErrCantDropFieldOrKey is strange here.

File: ddl/delete_range.go [hunk 5]
```diff
@@ -94,7 +94,15 @@ func (dr *delRange) addDelRangeJob(ctx context.Context, job *model.Job) error {
 	}
 	defer dr.sessPool.put(sctx)
 
-	err = insertJobIntoDeleteRangeTable(ctx, sctx, job)
+	if job.MultiSchemaInfo != nil {
+		var added bool
+		added, err = insertJobIntoDeleteRangeTableMultiSchema(ctx, sctx, job)
```
Comment (wjhuang2016) [inline ddl/delete_range.go:99]: `added == false` implies that `err is nil`, it may be changed later.

File: ddl/column.go (hunk 1 above)
Comment (tangenta) [reply ddl/column.go:667]: `ifExists` is the **last argument** in `DropColumn()`. As you said:

```
// indexIds is used to make sure we don't truncate args when decoding the rawArgs
```

It is impossible to have `indexIds` in old-version non-owner TiDB.

File: ddl/column.go (hunk 3 above)
Comment (tangenta) [reply ddl/column.go:578]: It is used to ignore the `ifExists` warning. If the column exists, `ErrCanDropFieldOrKey` is the only error can be reported. For details, see `checkDropColumn`.

File: ddl/ddl_worker.go (hunk 4 above)
Comment (tangenta) [reply ddl/ddl_worker.go:437]: Please take a look at the comment, it is not strange at all :)

https://github.com/pingcap/tidb/blob/ab27d4918a5a07d565ce3ef98761b6de6e90c9cc/util/dbterror/ddl_terror.go#L193-L194

File: ddl/delete_range.go (hunk 5 above)
Comment (tangenta) [reply ddl/delete_range.go]: Updated.

File: ddl/delete_range_util.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+package ddl
+
+type tblIdxID struct {
+	physicalID int64
+	indexID    int64
+}
+
+type elementIDAlloc struct {
+	physicalIDs map[int64]int64
```
Comment (wjhuang2016) [inline ddl/delete_range_util.go:23]: Why need a map?

File: ddl/ddl_worker.go [hunk 7]
```diff
@@ -444,6 +449,15 @@ func jobNeedGC(job *model.Job) bool {
 		case model.ActionDropSchema, model.ActionDropTable, model.ActionTruncateTable, model.ActionDropIndex, model.ActionDropPrimaryKey,
 			model.ActionDropTablePartition, model.ActionTruncateTablePartition, model.ActionDropColumn, model.ActionDropColumns, model.ActionModifyColumn, model.ActionDropIndexes:
 			return true
+		case model.ActionMultiSchemaChange:
```
Comment (wjhuang2016) [inline ddl/ddl_worker.go:452]: Can we add a UT for it?

File: ddl/delete_range_util.go (hunk 6 above)
Comment (tangenta) [reply ddl/delete_range_util.go:23]: To deduplicate the index IDs.

File: ddl/ddl_worker.go (hunk 7 above)
Comment (tangenta) [reply ddl/ddl_worker.go:455]: Updated.

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -3019,7 +3030,7 @@ func (d *ddl) AlterTable(ctx context.Context, sctx sessionctx.Context, ident ast
 		case ast.AlterTableAddColumns:
 			useMultiSchemaChange = true
 		case ast.AlterTableDropColumn:
-			err = d.DropColumns(sctx, ident, validSpecs)
+			useMultiSchemaChange = true
```
Comment (zimulala) [inline ddl/ddl_api.go:3033]:
```
case ast.AlterTableDropColumn, ast.AlterTableDropColumn:
         useMultiSchemaChange = true
```

File: ddl/ddl_worker.go [hunk 9]
```diff
@@ -432,8 +432,16 @@ func (w *worker) deleteRange(ctx context.Context, job *model.Job) error {
 	return errors.Trace(err)
 }
 
+// JobNeedGCForTest is only used for test.
+var JobNeedGCForTest = jobNeedGC
```
Comment (zimulala) [inline ddl/ddl_worker.go:436]: Could we put it to ddl_test.go？

Comment (tangenta) [reply ddl/ddl_worker.go]: Sure.

File: ddl/ddl_api.go (hunk 8 above)
Comment (tangenta) [reply ddl/ddl_api.go]: Updated.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**